### PR TITLE
Replace unwraps with expects in bevy_asset

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

- Panics in bevy_asset are often particularly frustrating and obtuse.

## Solution

- Provide somewhat more helpful error messages by replacing expects with unwraps.

## Meta

This is a very basic, mechanical pass to improve error handling.

I don't currently have a deep understanding of the internals (and they're likely to change), so the error messages provided are very localized. It would be nice to capture more information about the state in these expect messages as well before merging.

There are a few wasm-specific unwraps left in `wasm_asset_io.rs` that I didn't feel comfortable tackling. There are also some expects in test code left, which I felt were fine to keep.

Like always, improvements are very welcome.

### CODE REVIEW

File: crates/bevy_asset/src/asset_server.rs [hunk 1]
```diff
@@ -358,7 +362,7 @@ impl AssetServer {
         self.server
             .asset_io
             .watch_path_for_changes(asset_path.path())
-            .unwrap();
+            .expect("Could not watch path for changes.");
```
Comment (rparrett) [inline crates/bevy_asset/src/asset_server.rs:365]: Seems like this one could be `.watch_path_for_changes(asset_path.path())?;`

File: crates/bevy_asset/src/asset_server.rs [hunk 2]
```diff
@@ -399,7 +403,9 @@ impl AssetServer {
         let path = path.as_ref();
         if !self.server.asset_io.is_directory(path) {
             return Err(AssetServerError::AssetFolderNotADirectory(
-                path.to_str().unwrap().to_string(),
+                path.to_str()
+                    .expect("Could not convert path to string.")
```
Comment (rparrett) [inline crates/bevy_asset/src/asset_server.rs:407]: Seems like this should only fail if the path has invalid utf8.

**Review by rparrett — COMMENTED** [review]
Not informed enough to comment on a lot of these. But it seems like in many cases there's some type of `Error` being ignored that likely contains some useful information about what went wrong.

**Comment by alice-i-cecile** [comment]
>But it seems like in many cases there's some type of Error being ignored that likely contains some useful information about what went wrong.

Yep; that's my next pass for this PR.

File: crates/bevy_asset/src/asset_server.rs [hunk 3]
```diff
@@ -500,11 +506,13 @@ impl AssetServer {
     // triggered unless the `Assets` collection is actually updated.
     pub(crate) fn update_asset_storage<T: Asset>(&self, mut assets: ResMut<Assets<T>>) {
         let asset_lifecycles = self.server.asset_lifecycles.read();
-        let asset_lifecycle = asset_lifecycles.get(&T::TYPE_UUID).unwrap();
+        let asset_lifecycle = asset_lifecycles
+            .get(&T::TYPE_UUID)
+            .expect("Could not find an asset of the appropriate type UUID.");
```
Comment (IceSentry) [inline crates/bevy_asset/src/asset_server.rs:511]:
```suggestion
        let asset_lifecycle = asset_lifecycles.get(&T::TYPE_UUID).expect_or_else(|| {
            format!(
                "Could not find an asset of the appropriate type UUID={}.",
                T::TYPE_UUID
            )
        });
```

**Comment by IceSentry** [comment]
Sorry, my mistake, the method is unwrap_or_else() not expect_or_else. My local setup is giving me some issues and I didn't notice the mistake.

File: crates/bevy_asset/src/asset_server.rs [hunk 4]
```diff
@@ -500,11 +506,16 @@ impl AssetServer {
     // triggered unless the `Assets` collection is actually updated.
     pub(crate) fn update_asset_storage<T: Asset>(&self, mut assets: ResMut<Assets<T>>) {
         let asset_lifecycles = self.server.asset_lifecycles.read();
-        let asset_lifecycle = asset_lifecycles.get(&T::TYPE_UUID).unwrap();
+        let asset_lifecycle = asset_lifecycles.get(&T::TYPE_UUID).expect_or_else(|| {
+            format!(
+                "Could not find an asset of the appropriate type UUID={}.",
+                T::TYPE_UUID
+            )
+        });
```
Comment (IceSentry) [inline crates/bevy_asset/src/asset_server.rs:514]:
```suggestion
        let asset_lifecycle = asset_lifecycles.get(&T::TYPE_UUID).unwrap_or_else(|| {
            panic!(
                "Could not find an asset of the appropriate type UUID={}.",
                T::TYPE_UUID
            )
        });
```

Here's the actual line that compiles

**Comment by alice-i-cecile** [comment]
Closing per discussion in https://github.com/bevyengine/bevy/issues/3899.

**Pull request closed without merging**